### PR TITLE
Make vertical buttons translatable

### PR DESCRIPTION
**Pull request by zenseii**
This makes the vertical DISMISS and EXIT buttons in the hero screen translatable.

image

### CODE REVIEW

File: src/fheroes2/agg/agg_image.cpp [hunk 1]
```diff
@@ -1794,6 +1845,66 @@ namespace fheroes2
...
+                    // so we use our empty vertical button as a template
+                    _icnVsSprite[id][0] = GetICN( ICN::EMPTY_VERTICAL_GOOD_BUTTON, 0 );
+                    _icnVsSprite[id][1] = GetICN( ICN::EMPTY_VERTICAL_GOOD_BUTTON, 1 );
+
+                    // Copy the EXIT text back from the original button
+                    const Sprite & originalReleased = GetICN( ICN::HSBTNS, 0 );
+                    const Sprite & originalPressed = GetICN( ICN::HSBTNS, 1 );
+
+                    Copy( originalReleased, 9, 2, _icnVsSprite[id][0], 4, 2, 21, 112 );
+                    Copy( originalPressed, 9, 5, _icnVsSprite[id][1], 3, 5, 19, 111 );
```
Comment (Districh-ru) [inline src/fheroes2/agg/agg_image.cpp:1892]: For original EXIT button we need to use 2 and 3 image index and use different X offsets in these images when performing `Copy()`

```suggestion
                    const Sprite & originalReleased = GetICN( ICN::HSBTNS, 2 );
                    const Sprite & originalPressed = GetICN( ICN::HSBTNS, 3 );

                    Copy( originalReleased, 4, 2, _icnVsSprite[id][0], 4, 2, 21, 112 );
                    Copy( originalPressed, 3, 5, _icnVsSprite[id][1], 3, 5, 19, 111 );
```

File: src/fheroes2/agg/agg_image.cpp [hunk 2]
```diff
@@ -3965,6 +4052,70 @@ namespace fheroes2
...
+                    FillTransform( released, 0, 1, 3, 1, 1 );
+                    FillTransform( released, 0, 2, 2, 1, 1 );
+                    FillTransform( released, 0, 3, 1, 1, 1 );
+
+                    FillTransform( released, released.width() - 2, 0, 2, 1, 1 );
+                    FillTransform( released, released.width() - 1, 1, 1, 1, 1 );
+
+                    FillTransform( released, released.width() - 1, released.height() - 3, 1, 3, 1 );
+                    FillTransform( released, released.width() - 2, released.height() - 2, 1, 2, 1 );
+                    FillTransform( released, released.width() - 3, released.height() - 1, 1, 1, 1 );
```
Comment (Districh-ru) [inline src/fheroes2/agg/agg_image.cpp:4114]: We can use `setButtonCornersTransparent( released );` here. It does the same for the released button. :)
In future we can also modify this function to make transparent corners for the pressed buttons (we can leave a TODO).

File: src/fheroes2/gui/ui_button.cpp [hunk 3]
```diff
@@ -642,7 +648,7 @@ namespace fheroes2
 
         const fheroes2::Text releasedText( supportedText, releasedButtonFont );
         const fheroes2::Text pressedText( supportedText, { fheroes2::FontSize::BUTTON_PRESSED, buttonFont } );
-        const int32_t textWidth = releasedText.width();
+        const int32_t textWidth = releasedText.width( releasedText.width() );
```
Comment (Districh-ru) [inline src/fheroes2/gui/ui_button.cpp:651]: In this line we are trying to get the width of `releasedText`, which is limited to the width of the same `releasedText`.
Maybe there is a typo here? :)

**Review by Districh-ru — CHANGES_REQUESTED** [review]
Hi, @zenseii, I left several comments to the code. Could you please take a look in your spare time.

Also comparing original buttons to the new ones I noticed that the text vertical spacing is 1 pixel smaller in original buttons:
new (16px) <-> original (15px)
изображение <->  изображение

Analyzing the main menu buttons I found the same difference:
изображение <-> изображение


Maybe we should change the return value to 15 here:
https://github.com/ihhub/fheroes2/blob/edd2ca7b08b5b89b4c7cafa28b969e6e8aa90a91/src/fheroes2/gui/ui_text.cpp#L483

What do you think?

File: src/fheroes2/gui/ui_button.cpp (hunk 3 above)
Comment (Districh-ru) [reply src/fheroes2/gui/ui_button.cpp:651]: We can currently use the `maximumTextAreaWidth` variable here (`const int32_t textWidth = releasedText.width( maximumTextAreaWidth );`) not to call the `.width()` method.

Comment (Districh-ru) [reply src/fheroes2/gui/ui_button.cpp:651]: Could you please add a TODO above `width()` and `height()` methods in ui_text.cpp (lines 509 and 514). Something like: "Properly handle strings with many text lines ('\n')".

File: src/fheroes2/agg/agg_image.cpp [hunk 4]
```diff
@@ -347,6 +349,47 @@ namespace
 
     OriginalAlphabetPreserver alphabetPreserver;
 
+    // This class is used for situations when we need to remove letter-specific offsets, like when we display single letters in a row,
+    // and then restore these offsets within the scope of the code
+    class ButtonFontRestorer
```
Comment (ihhub) [inline src/fheroes2/agg/agg_image.cpp:354]: To be more accurate would it better to name this class as **ButtonFontOffsetRestorer**?

File: src/fheroes2/agg/agg_image.cpp [hunk 5]
```diff
@@ -347,6 +349,47 @@ namespace
...
+
+    private:
+        std::vector<fheroes2::Sprite> & _font;
+        std::vector<int32_t> _originalXOffsets;
+    };
+
+    ButtonFontRestorer::ButtonFontRestorer( std::vector<fheroes2::Sprite> & font, const int32_t offsetX )
+        : _font( font )
+    {
+        _originalXOffsets.reserve( _font.size() );
```
Comment (ihhub) [inline src/fheroes2/agg/agg_image.cpp:372]: Just a very tiny nitpicking: here we use **_font** variable but a line lower we use **font**. It's better to be persistent :)

**Review by ihhub — APPROVED** [review]
Hi @zenseii , everything looks good to me. I left 2 optional comments. Please ignore them if you think they are useless.

**Review by Districh-ru — APPROVED** [review]
Well done, @zenseii !

**Review by Districh-ru — APPROVED** [review]
👍

**Review by ihhub — APPROVED** [review]

**Comment by ihhub** [comment]
@zenseii , great work!

**Pull request merged**